code/crates/test/tests/it/equivocation.rs

```diff
@@ -0,0 +1,61 @@
+use std::time::Duration;
+
+use informalsystems_malachitebft_test::{Address, Height, Proposal, Value};
+use malachitebft_core_consensus::LocallyProposedValue;
+use malachitebft_core_types::{Round, SignedProposal};
+use malachitebft_engine::{consensus::Msg, network::NetworkEvent};
+use malachitebft_peer::PeerId;
+use malachitebft_signing_ed25519::Signature;
+use malachitebft_test_framework::TestParams;
+
+use crate::TestBuilder;
+
+#[tokio::test]
+pub async fn equivocation_proposer() {
```

Comment:
@romac, the inject feature is not enough/complete. Indeed, injecting a `ProposeValue` message does not work as the engine/driver will not consider two propose values; therefore, no conflicting proposals are sent. If we want to force receiving a proposal via a `NetworkEvent`, it also does not work since we do not have access to the peer id and cannot sign the proposal.
Do you have an idea on how to fix this?

Reply:
Ah bummer… Let me think about it and I'll come back to you

Reply:
Maybe we can pass an unsigned message to `inject` and use `Node::get_signing_provided` to sign the message in the test framework?

Reply:
For that, we need to have access to the [app instance](https://github.com/informalsystems/malachite/blob/main/code/crates/test/app/src/node.rs#L55) (on which we can actually call `get_signing_provider`), but I don't see where we can access it in the framework.

Reply:
Maybe an overkill but can you run once and collect the proposals from that run? And then run again (with same keys) and inject in the second run the proposal from the first?

Reply:
The other question, does it make sense to do the tests in a separate PR and merge this one as is?

Reply:
> Do you mean hardcoding the messages? For example, for votes, retrieve data from such logs:

I thought you could intercept the proposal (or vote) in the first run, save it in some test state and then inject it in the second run.

> I would likely prefer the inject approach, which might also be useful for other tests.

You still do the inject but you need a properly signed (second) proposal

Reply:
Nifty idea, I'll give it a go next week!

Reply:
Nifty idea, I'll give it a go next week!